File: lib/agent.js
import axios from 'axios';
import { getSystemContext } from './utils.js';

export class OllamaCodeAgent {
    constructor(settings) {
        this.settings = settings;
        this.history = [];
        this.groqUrl = "https://api.groq.com/openai/v1/chat/completions";
    }

    clipHistory(maxWords = 4000) {
        let totalWords = 0;
        let clipped = [];
        for (let i = this.history.length - 1; i >= 0; i--) {
            const msg = this.history[i];
            const words = msg.content.split(/\s+/).length;
            if (totalWords + words > maxWords) break;
            clipped.unshift(msg);
            totalWords += words;
        }
        this.history = clipped;
    }

    async askAi(cwd) {
        const { listInstalledPlugins } = await import('./utils.js');
        const plugins = await listInstalledPlugins();
        const pluginInfo = plugins.length > 0 ? `Installed Plugins: ${plugins.join(', ')}` : "No custom plugins installed.";
        
        const systemMsg = `
# Role: OllamaCode v1.2.1
You are an expert autonomous terminal agent. Your goal is to help the user with coding, debugging, and system administration tasks directly from the shell.

## Context:
- Working Directory: ${cwd}
- System Info: ${getSystemContext(cwd)}
- ${pluginInfo}

## Capabilities:
1. **Explore**: Use \`ollamacode tree\` to see the project structure.
2. **Analyze**: Use \`ollamacode cat-file <file_path>\` to read files with line numbers.
3. **Execute**: Suggest shell commands in \`\`\`bash ... \`\`\` blocks.
4. **Modify**: Use \`ollamacode write-file <file_path> <content>\` for surgical edits.
5. **Plugins**: Run custom tools via \`ollamacode run <plugin_name> [args]\`.

## Guidelines:
- **Action-First**: NEVER just print code. You MUST write it to a file using \`ollamacode write-file\`.
- **Command Blocks**: All commands MUST be inside \`\`\`bash ... \`\`\` blocks to be executed.
- **Step-by-Step**: If a folder is requested, 1. \`mkdir <name>\`, 2. \`cd <name>\`, 3. \`ollamacode write-file ...\`.
- **Precision**: Use the exact folder/file names requested by the user.
- **Be Brief**: Minimize talk. Execute first, explain only if necessary.
- If the task is finished, state "Task complete."
`.trim();
        this.clipHistory();
        const messages = [{ role: "system", content: systemMsg }, ...this.history];

        try {
            if (this.settings.provider === "Groq") {
                const response = await axios.post(this.groqUrl, {
                    messages,
                    model: this.settings.model
                }, {
                    headers: {
                        Authorization: `Bearer ${this.settings.api_key}`,
                        'Content-Type': 'application/json'
                    },
                    timeout: 30000
                });
                let content = response.data.choices[0].message.content;
                let filtered = content.replace(/<think>[\s\S]*?<\/think>/gi, '').trim();
                return filtered || "I have analyzed the situation. Let me suggest a command.";
            } else {
                const response = await axios.post(`${this.settings.ollama_url}/api/chat`, {
                    messages,
                    model: this.settings.model,
                    stream: false
                }, { timeout: 150000 });
                let content = response.data.message.content;
                let filtered = content.replace(/<think>[\s\S]*?<\/think>/gi, '').trim();
                return filtered || "I have analyzed the situation. Let me suggest a command.";
            }
        } catch (e) {
            return `Connection Error: ${e.response?.data?.error?.message || e.message}`;
        }
    }

    getCommands(answer) {
        const matches = [...answer.matchAll(/```(?:bash|sh|shell)\n([\s\S]*?)```/gi)];
        return matches.map(m => m[1].trim());
    }
}

File: ollamacode/agent.py
import requests
import re
from .utils import get_system_context

class OllamaCodeAgent:
    def __init__(self, settings):
        self.settings = settings
        self.history = []
        self.groq_url = "https://api.groq.com/openai/v1/chat/completions"

    def clip_history(self, max_words=4000):
        total_words = 0
        clipped = []
        for msg in reversed(self.history):
            words = len(msg['content'].split())
            if total_words + words > max_words: break
            clipped.insert(0, msg)
            total_words += words
        self.history = clipped

    def ask_ai(self, cwd):
        from .utils import list_installed_plugins
        plugins = list_installed_plugins()
        plugin_info = f"Installed Plugins: {', '.join(plugins)}" if plugins else "No custom plugins installed."
        
        system_msg = (
            "# Role: OllamaCode v1.2.1\n"
            "You are an expert autonomous terminal agent. Your goal is to help the user with coding, debugging, and system administration tasks directly from the shell.\n\n"
            "## Context:\n"
            f"- Working Directory: {cwd}\n"
            f"- System Info: {get_system_context(cwd)}\n"
            f"- {plugin_info}\n\n"
            "## Capabilities:\n"
            "1. **Explore**: Use `ollamacode tree` to see the project structure.\n"
            "2. **Analyze**: Use `ollamacode cat-file <file_path>` to read files with line numbers.\n"
            "3. **Execute**: Suggest shell commands in ```bash ... ``` blocks.\n"
            "4. **Modify**: Use `ollamacode write-file <file_path> <content>` for surgical edits.\n"
            "5. **Plugins**: Run custom tools via `ollamacode run <plugin_name> [args]`.\n\n"
            "## Guidelines:\n"
            "- **Action-First**: NEVER just print code snippets. You MUST write them to files using `ollamacode write-file`.\n"
            "- **Command Blocks**: All actions MUST be wrapped in ```bash ... ``` blocks for automatic execution.\n"
            "- **Chaining**: Create directories with `mkdir`, enter them with `cd`, and then write files.\n"
            "- **Precision**: Strictly follow the user's requested names (folders/files).\n"
            "- **Be Brief**: Minimize explanations. Focus on executing the task.\n"
            "- If the task is finished, state \"Task complete.\""
        )
        self.clip_history()
        messages = [{"role": "system", "content": system_msg}] + self.history

        try:
            if self.settings["provider"] == "Groq":
                headers = {"Authorization": f"Bearer {self.settings['api_key']}", "Content-Type": "application/json"}
                response = requests.post(self.groq_url, headers=headers, json={"messages": messages, "model": self.settings["model"]}, timeout=30)
                if response.status_code == 200:
                    content = response.json()['choices'][0]['message']['content']
                    filtered = re.sub(r'<think>.*?</think>', '', content, flags=re.DOTALL).strip()
                    return filtered if filtered else "I have analyzed the situation. Let me suggest a command."
            else:
                response = requests.post(f"{self.settings['ollama_url']}/api/chat", json={"model": self.settings["model"], "messages": messages, "stream": False}, timeout=150)
                if response.status_code == 200:
                    content = response.json()['message']['content']
                    filtered = re.sub(r'<think>.*?</think>', '', content, flags=re.DOTALL).strip()
                    return filtered if filtered else "I have analyzed the situation. Let me suggest a command."
            return f"Error: {response.status_code}"
        except Exception as e: return f"Connection Error: {str(e)}"

    def get_commands(self, answer):
        return re.findall(r'```(?:bash|sh|shell)\n(.*?)```', str(answer), re.DOTALL | re.IGNORECASE)
